drivers/i2s/i2s_nrfx_tdm.c

```diff
@@ -0,0 +1,1143 @@
+	tdm_ctrl_t *ctrl_data = drv_data->drv_cfg->control_data;
+	NRF_TDM_Type *p_reg = drv_data->drv_cfg->p_reg;
+
+	nrf_tdm_configure(p_reg, p_config);
+	nrf_tdm_mck_set(p_reg, p_config->mck_setup != 0);
+
+	ctrl_data->handler = handler;
+
+	nrf_tdm_event_clear(p_reg, NRF_TDM_EVENT_RXPTRUPD);
+	nrf_tdm_event_clear(p_reg, NRF_TDM_EVENT_TXPTRUPD);
+	nrf_tdm_event_clear(p_reg, NRF_TDM_EVENT_STOPPED);
+	NRFX_IRQ_ENABLE(nrfx_get_irq_number(p_reg));
+}
+
+static void clock_started_callback(struct onoff_manager *mgr, struct onoff_client *cli,
+				   uint32_t state, int res)
+{
+	struct tdm_drv_data *drv_data = CONTAINER_OF(cli, struct tdm_drv_data, clk_cli);
+
+	/* The driver state can be set back to READY at this point if the DROP
+	 * command was triggered before the clock has started. Do not start
+	 * the actual transfer in such case.
+	 */
+	if (drv_data->state == I2S_STATE_READY) {
+		tdm_uninit(drv_data);
+		(void)audio_clock_release();
+	} else {
+		(void)start_transfer(drv_data);
+	}
+}
+
+static int trigger_start(const struct device *dev)
+{
+	struct tdm_drv_data *drv_data = dev->data;
+	const struct tdm_drv_cfg *drv_cfg = dev->config;
+	int ret;
+	const nrf_tdm_config_t *nrfx_cfg = (drv_data->active_dir == I2S_DIR_TX)
+						   ? &drv_data->tx.nrfx_cfg
+						   : &drv_data->rx.nrfx_cfg;
+
+	tdm_init(drv_data, nrfx_cfg, drv_cfg->data_handler);
+
+	drv_data->state = I2S_STATE_RUNNING;
+
+	nrf_tdm_sck_configure(drv_cfg->p_reg,
+			      drv_cfg->clk_src == ACLK ? NRF_TDM_SRC_ACLK : NRF_TDM_SRC_PCLK32M,
+			      false);
+
+	nrf_tdm_mck_configure(drv_cfg->p_reg,
+			      drv_cfg->clk_src == ACLK ? NRF_TDM_SRC_ACLK : NRF_TDM_SRC_PCLK32M,
+			      false);
+	/* If it is required to use certain HF clock, request it to be running
+	 * first. If not, start the transfer directly.
+	 */
+	if (drv_data->request_clock) {
+		sys_notify_init_callback(&drv_data->clk_cli.notify, clock_started_callback);
+		ret = audio_clock_request(drv_data);
+		if (ret < 0) {
+			tdm_uninit(drv_data);
+			drv_data->state = I2S_STATE_READY;
+
+			LOG_ERR("Failed to request clock: %d", ret);
+			return -EIO;
+		}
+	} else {
+		ret = start_transfer(drv_data);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int tdm_nrfx_trigger(const struct device *dev, enum i2s_dir dir, enum i2s_trigger_cmd cmd)
+{
+	struct tdm_drv_data *drv_data = dev->data;
+	const struct tdm_drv_cfg *drv_cfg = dev->config;
+	bool configured = false;
+	bool cmd_allowed;
+
+	/* This driver does not use the I2S_STATE_NOT_READY value.
+	 * Instead, if a given stream is not configured, the respective
+	 * flag (tx_configured or rx_configured) is cleared.
+	 */
+	channels_configuration_merge(&drv_data->tx.nrfx_cfg.channels,
+				     &drv_data->rx.nrfx_cfg.channels);
+	if (dir == I2S_DIR_BOTH) {
+		if (!channels_configuration_check(drv_data->tx.nrfx_cfg.channels,
+						  drv_data->rx.nrfx_cfg.channels)) {
+			LOG_ERR("TX and RX channels configurations are different");
+			return -EIO;
+		}
```

Comment:
This should not be done here, but a few lines below, where configurations for TX and RX are compared (in the current form, the comment right above is quite unclear).